enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,298 @@ package hit_tracker_client
+		}
+		return
+	}
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	groupID := h.groupID(ctx)
+	if _, ok := h.hitsByGroup[groupID]; !ok {
+		h.hitsByGroup[groupID] = &cacheHits{
+			hits: make([]*hitpb.CacheHit, 0, initialSliceCapacity),
+		}
+	}
+
+	groupHits := h.hitsByGroup[groupID]
+	if len(groupHits.hits) >= h.maxPendingHitsPerGroup {
+		log.CtxWarning(ctx, "Exceeded maximum number of pending cache hits, dropping some.")
+		metrics.RemoteHitTrackerUpdates.With(
+			prometheus.Labels{
+				metrics.GroupID:              string(groupID),
+				metrics.EnqueueUpdateOutcome: "dropped_too_many_updates",
+			}).Add(float64(1))
+		return
+	}
+	metrics.RemoteHitTrackerUpdates.With(
+		prometheus.Labels{
+			metrics.GroupID:              string(groupID),
+			metrics.EnqueueUpdateOutcome: "enqueued",
+		}).Add(float64(1))
+
+	authHeaders := authutil.GetAuthHeaders(ctx, h.authenticator)
+	groupHits.authHeaders = authHeaders
+	groupHits.hits = append(groupHits.hits, hit)
+}
+
 type TransferTimer struct {
+	ctx             context.Context
+	enqueueFn       func(context.Context, *repb.RequestMetadata, *hitpb.CacheHit)
+	invocationID    string
+	requestMetadata *repb.RequestMetadata
+	digest          *repb.Digest
+	start           time.Time
+	client          hitpb.HitTrackerServiceClient
 }
 
 func (t *TransferTimer) CloseWithBytesTransferred(bytesTransferredCache, bytesTransferredClient int64, compressor repb.Compressor_Value, serverLabel string) error {
+	hit := &hitpb.CacheHit{
+		RequestMetadata: t.requestMetadata,
+		Resource: &rspb.ResourceName{
+			Digest:    t.digest,
+			CacheType: rspb.CacheType_CAS,
+		},
+		SizeBytes: bytesTransferredClient,
+		Duration:  durationpb.New(time.Since(t.start)),
+	}
+	t.enqueueFn(t.ctx, t.requestMetadata, hit)
 	return nil
 }
 
 func (t *TransferTimer) Record(bytesTransferred int64, duration time.Duration, compressor repb.Compressor_Value) error {
-	return nil
+	return status.InternalError("Unxpected call to hit_tracker_client.Record()")
+}
+
+func (h *HitTrackerClient) TrackDownload(digest *repb.Digest) interfaces.TransferTimer {
+	return &TransferTimer{
+		ctx:             h.ctx,
+		enqueueFn:       h.enqueueFn,
+		requestMetadata: h.requestMetadata,
+		digest:          digest,
+		start:           time.Now(),
+		client:          h.client,
+	}
 }
 
-func (h *HitTracker) TrackDownload(d *repb.Digest) interfaces.TransferTimer {
-	return &TransferTimer{}
+// Writes hit the backing cache, so no need to report on hit-tracking.
+func (h *HitTrackerClient) TrackUpload(digest *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
 }
 
-func (h *HitTracker) TrackUpload(d *repb.Digest) interfaces.TransferTimer {
-	return &TransferTimer{}
+func (h *HitTrackerFactory) runWorker() {
+	for {
+		select {
+		case <-h.quit:
+			h.stopped.Store(false)
+			return
+		case <-h.ticker:
+			h.sendTrackRequests(context.Background())
+		}
+	}
+}
+
+func (h *HitTrackerFactory) sendTrackRequests(ctx context.Context) int {
```

Comment:
nit: comment what the returned int means

Reply:
Done